src/groups.jl

```diff
@@ -23,16 +23,26 @@ function addgroup!(suite::BenchmarkGroup, id, args...)
     return g
 end
 
+clear_empty!(x) = x
+function clear_empty!(group::BenchmarkGroup)
+    for (k,v) in pairs(group)
+        if v isa BenchmarkGroup && isempty(v)
+            delete!(group, k)
+        end
+    end
+    group
+end
+
 # Dict-like methods #
 #-------------------#
 
 Base.:(==)(a::BenchmarkGroup, b::BenchmarkGroup) = a.tags == b.tags && a.data == b.data
 Base.copy(group::BenchmarkGroup) = BenchmarkGroup(copy(group.tags), copy(group.data))
 Base.similar(group::BenchmarkGroup) = BenchmarkGroup(copy(group.tags), empty(group.data))
-Base.isempty(group::BenchmarkGroup) = isempty(group.data)
+Base.isempty(group::BenchmarkGroup) = isempty(clear_empty!(group).data)
```

Comment:
I don't like the fact that this is implicitly mutating

Reply:
I think because subgroups are now created automatically upon access, pruning them eagerly like this doesn’t feel so much like a “mutation” (though it technically is), but more like a reduction to a standard form. ie

```julia
suite[1] = BenchmarkGroup()
isempty(suite[1])
suite[1][2] = ...
```

which works the exact same as before, due to this automatic creation.

This is used as a part of `clear_empty!` — you first need to clear out all empty children before seeing whether a parent `BenchmarkGroup()` is also empty. (Borrowed this approach from EasyConfig.jl.) This could be without changing the BenchmarkGroup(), but it would result in a recursive blow-up, because every time you call `isempty` inside `clear_empty!`, it would be creating copies of the BenchmarkGroup — rather than just working on a single copy and cleaning it up as it goes.

Reply:
Ok I understand! Maybe it would make sense to document it in a docstring, along with other key functions from Base, and add said docstrings to the API reference page?

Reply:
Done